### ISSUE TITLE
source-foreign: UX improvements

### ISSUE THREAD

**Issue by FlyingWombat** [xonsh/xonsh#5318]
I'm just  trying out xonsh, and would like to discuss an improvement for UX.

Sourcing foreign shell functions is an incomplete feature (see #1056 , #5043  ), and so should be disabled by default. It should be feature-gated behind something like an `--experimental` flag so that it is clear to users that it likely won't work as they expect. 

The source-foreign command, as far as I can tell, is intended to ease the transition to xonsh for new users and provide integration with existing system configurations targeting bash (e.g. /etc/profile). 
There are still a significant portion of bash functions that won't work when imported, so in both cases it is important for users to know they need to check the foreign function imports on a case-by-case basis.

Foreign shell functions, in their current state, are a stumbling block that give the impression that xonsh is buggy.

----------------------------

Additionally, source-foreign `--non-interactive` flag should be the default. It is unexpected behavior for things to be imported that aren't defined in the script being imported.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by Qyriad** [comment]
I strongly disagree that `source-foreign` should be disabled by default or feature gated, and I don't agree that it's best used for transitioning new users to Xonsh — in fact I think that's probably its weakest use case. I strongly believe that source-foreign is one of Xonsh's best features and one of the reasons I can wholeheartedly recommend Xonsh — I do not fear for the inevitable damn cases where all I have to setup something complex is some bespoke shell script to source, and that's not even counting the convenience of parsing commands that expect to be sh-sourced like `tmux show-environment -s` or `ssh-agent -s`.

I also don't understand your impression that it's extremely buggy — I've certainly run into things that don't work, but it's good enough to use the *rather* complex machinery of [Nix](https://github.com/NixOS/nix)'s `print-dev-env` command (which outputs a bash script), and in general I've found Xonsh's `source-bash` to work far better than some other compatibility equivalents in other exotic shells. Maybe it would help if you could give some concrete examples that don't work (besides the linked issues).

**Comment by FlyingWombat** [comment]
tl;dr I'm just saying ForeignShellFuctionAlias should be put behind e.g. an `--experimental` flag.

You seem to misunderstand what I'm proposing here. I'm not suggesting disabling or getting rid of `source-foreign` -- only that the *incomplete* parts of it, such as `ForeignShellFunctionAlias` be marked experimental and require a flag, such as `--experimental`, to be used. The linked issues, and possibly others I haven't run into, mean that there is a good chance ForeignShellFunctionAlias will not work as expected. It is poor UX because users expect it to do what bash does; it needs to be clear that it does not, and for which cases it does not. In other words, the issue isn't that source-foreign is buggy, but that it doesn't work as expected. For me, some functions failed silently (because of the linked issues) and I spent a couple hours trying to figure out if my knowledge of xonsh is bad (since I just started), or if xonsh has a bug. I want to spare other users from the same headache.

A way to improve UX, that I'm proposing, is to have source-foreign ignore function definitions unless `--experimental` or `--functions`, or something like that is passed as an option. And, the same should be done for all other incomplete features.

P.S.
It would be even better to just fix the issues, and complete all the released features -- but considering #1056 has been open for 8 years now, I don't expect that will happen any time soon. In the mean-time it should be put behind a flag.

**Comment by gforsyth** [comment]
Thanks for the feedback @FlyingWombat.  Software is never complete, and `xonsh` will _never_ be complete.  We make an effort not to break users, but there are no promises.  
That said, gating existing functionality behind a new feature flag because something has been "incomplete" for 8 years doesn't seem like something we're going to do.  I'm going to close this out as unplanned, but happy to continue the conversation.

**Comment by FlyingWombat** [comment]
@gforsyth  I understand the reason for not wanting to gate a feature that has been around for a long time, since many users will be used to how it works (including its unexpected limitations) and will have to update their xonsh code to accommodate that change. At the very least, I think the documentation around source-foreign should explain these limitations, until they are addressed. 

>  Software is never complete, and xonsh will never be complete.

`hello-world` is complete. Software with *variable project scope* is never complete. Fixed scopes can be complete, and parts (feature-scopes) can be complete even if the whole is not. 

The point here isn't to debate when you can and can't add new or modify features to a project; the point is about setting user expectations. If users will naturally expect Y with X, but Y is not with X in the current release, then they need to know about that. That is the crux of this issue.

An intuitive way to communicate to users that a feature is released in an incomplete state is feature-gating -- which is why I proposed that. In this case, documentation or a help message may suffice.

---------------

I still think the second proposal, making `--non-interactive` the default behavior, is a UX improvement for reducing unexpected behavior. The current behavior can also lead to issues, such as clobbering an environment variables or aliases that aren't defined in the target shell script (e.g. unexpectedly changing PATH, because some system shell script got pulled in). But if you think that it is "too late" to make that change (e.g. too many users rely on the current behavior) that is an acceptable reason to reject it -- and if so, documentation should be updated.

**Comment by gforsyth** [comment]
PRs for improving the docs are very welcome.
